internal/dataplane/kongstate/kongstate.go

```diff
@@ -718,29 +718,64 @@ func (ks *KongState) FillCustomEntities(
 			continue
 		}
 		// Unmarshal fields of the entity.
-		var parsedEntity map[string]interface{}
+		var parsedEntity map[string]any
 		if err = json.Unmarshal(entity.Spec.Fields.Raw, &parsedEntity); err != nil {
 			failuresCollector.PushResourceFailure(fmt.Sprintf("failed to unmarshal fields of entity: %v", err), entity)
 			continue
 		}
 		// Fill the "foreign" fields if the entity has such fields referencing services/routes/consumers.
-		ks.fillCustomEntityForeignFields(logger, entity, schema, parsedEntity, pluginRels, workspace)
-		// Put the entity into the custom collection to store the entities of its type.
-		if _, ok := ks.CustomEntities[entity.Spec.EntityType]; !ok {
-			ks.CustomEntities[entity.Spec.EntityType] = &KongCustomEntityCollection{
-				Schema: schema,
+		// First Find out possible foreign field combinations attached to the KCE resource.
+		foreignFieldCombinations := ks.findCustomEntityForeignFields(logger, entity, schema, pluginRels, workspace)
+		// If there are any foreign fields, generate one entity per each foreign entity combination.
+		if len(foreignFieldCombinations) > 0 {
+			for _, combination := range foreignFieldCombinations {
+				// REVIEW: We do not have a built-in or common method to deep copy a `map[string]any`,
+				// so we decode it from the original JSON each time when generate an entity.
+				var parsedEntity map[string]any
+				if err = json.Unmarshal(entity.Spec.Fields.Raw, &parsedEntity); err != nil {
+					continue
+				}
+				generatedEntity := CustomEntity{
+					K8sKongCustomEntity: entity,
+					ForeignEntityIDs:    make(map[kong.EntityType]string),
+				}
+				// fill the fields referring to foreign entities.
+				for _, foreignField := range combination {
+					parsedEntity[foreignField.fieldName] = map[string]any{
+						"id": foreignField.foreignEntityID,
+					}
+					// save the referred foreign entity IDs for sorting.
+					generatedEntity.ForeignEntityIDs[foreignField.foreignEntityType] = foreignField.foreignEntityID
+				}
+				generatedEntity.Object = parsedEntity
+				ks.addCustomEntity(entity.Spec.EntityType, schema, generatedEntity)
 			}
+		} else {
+			ks.addCustomEntity(
+				entity.Spec.EntityType, schema,
+				CustomEntity{
+					Object:              parsedEntity,
+					K8sKongCustomEntity: entity,
+				},
+			)
 		}
-		collection := ks.CustomEntities[entity.Spec.EntityType]
-		collection.Entities = append(collection.Entities, CustomEntity{
-			Object:              parsedEntity,
-			K8sKongCustomEntity: entity,
-		})
 	}
 
 	ks.sortCustomEntities()
 }
```

Comment:
How about extracting the part responsible for generating custom entities for a KCE object to a separate function? Something like below (I can't suggest it via GH suggestion as there are lines removals):

```go

		// Fill the "foreign" fields if the entity has such fields referencing services/routes/consumers.
		// First, find out possible foreign field combinations attached to the KCE resource.
		foreignFieldCombinations := ks.findCustomEntityForeignFields(logger, entity, schema, pluginRels, workspace)
		// Then, based on the foreign field combinations, generate custom entities.
		customEntities, err := generateCustomEntities(entity, foreignFieldCombinations)
		if err != nil {
			failuresCollector.PushResourceFailure(fmt.Sprintf("failed to generate custom entities: %v", err), entity)
			continue
		}
		for _, e := range customEntities {
			ks.addCustomEntity(entity.Spec.EntityType, schema, e)
		}
	}

	ks.sortCustomEntities()
}

// generateCustomEntities generates custom entities based on the foreign field combinations.
func generateCustomEntities(entity *kongv1alpha1.KongCustomEntity, foreignFieldCombinations [][]entityForeignFieldValue) ([]CustomEntity, error) {
	copyEntityFields := func() (map[string]any, error) {
		// Unmarshal the fields of the entity to have a fresh copy for each combination as we may modify them.
		fields := map[string]any{}
		if err := json.Unmarshal(entity.Spec.Fields.Raw, &fields); err != nil {
			return nil, fmt.Errorf("failed to unmarshal entity fields: %w", err)
		}
		return fields, nil
	}

	// If there are any foreign fields, generate one entity per each foreign entity combination.
	if len(foreignFieldCombinations) > 0 {
		var customEntities []CustomEntity
		for _, combination := range foreignFieldCombinations {
			entityFields, err := copyEntityFields()
			if err != nil {
				return nil, err
			}
			generatedEntity := CustomEntity{
				K8sKongCustomEntity: entity,
				ForeignEntityIDs:    make(map[kong.EntityType]string),
				Object:              entityFields,
			}
			// Fill the fields referring to foreign entities.
			for _, foreignField := range combination {
				entityFields[foreignField.fieldName] = map[string]any{
					"id": foreignField.foreignEntityID,
				}
				// Save the referred foreign entity IDs for sorting.
				generatedEntity.ForeignEntityIDs[foreignField.foreignEntityType] = foreignField.foreignEntityID
			}
			customEntities = append(customEntities, generatedEntity)
		}
		return customEntities, nil
	}

	// Otherwise (no foreign fields), generate a single entity.
	entityFields, err := copyEntityFields()
	if err != nil {
		return nil, err
	}
	return []CustomEntity{{
		K8sKongCustomEntity: entity,
		Object:              entityFields,
	}}, nil
}
```

Reply:
extracted to `generateCustomEntities`.